src/main/java/mate/academy/dao/impl/UserDaoImpl.java

```diff
@@ -0,0 +1,60 @@
+package mate.academy.dao.impl;
+
+import jakarta.persistence.criteria.CriteriaQuery;
+import java.util.List;
+import java.util.Optional;
+import mate.academy.dao.UserDao;
+import mate.academy.exception.DataProcessingException;
+import mate.academy.lib.Dao;
+import mate.academy.model.User;
+import mate.academy.util.HibernateUtil;
+import org.hibernate.Session;
+import org.hibernate.Transaction;
+
+@Dao
+public class UserDaoImpl implements UserDao {
+    @Override
+    public User add(User user) {
+        Transaction transaction = null;
+        Session session = null;
+        try {
+            session = HibernateUtil.getSessionFactory().openSession();
+            transaction = session.beginTransaction();
+            session.persist(user);
+            transaction.commit();
+            return user;
+        } catch (Exception e) {
+            if (transaction != null) {
+                transaction.rollback();
+            }
+            throw new DataProcessingException("Can't insert new user " + user, e);
+        } finally {
+            if (session != null) {
+                session.close();
+            }
+        }
+    }
+
+    @Override
+    public Optional<User> getByEmail(String email) {
+        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
+            return session.createQuery("FROM User WHERE email = :email", User.class)
+                          .setParameter("email", email)
+                          .uniqueResultOptional();
+        } catch (Exception e) {
+            throw new DataProcessingException("Can't get a user by email: " + email, e);
+        }
+    }
+
+    @Override
+    public List<User> getAll() {
+        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
+            CriteriaQuery<User> criteriaQuery = session.getCriteriaBuilder()
+                                                       .createQuery(User.class);
+            criteriaQuery.from(User.class);
+            return session.createQuery(criteriaQuery).getResultList();
```

Comment:
You can simplify it:
```suggestion
            return session.createQuery("FROM User", User.class).getResultList();
```